hasher.go

```diff
@@ -415,3 +483,88 @@ func (h *Hasher) merkleizeImpl(dst []byte, input []byte, limit uint64) []byte {
 	}
 	return append(dst, res...)
 }
+
+func merkleizeInput(input []byte, limit uint64) []byte {
+	elemCount := len(input) / 32
+	elements := make([][32]byte, elemCount)
+	elemLen := len(elements)
+	for i, j := 0, 0; j < elemLen; i, j = i+32, j+1 {
+		if j == elemLen-1 {
+			copy(elements[j][:], input[i:])
+		} else {
+			copy(elements[j][:], input[i:i+32])
+		}
+	}
+
+	var result [32]byte
+	if limit == 0 {
+		result = merkleizeVector(elements, uint64(elemLen))
+	} else {
+		result = merkleizeVector(elements, limit)
+	}
+
+	return result[:]
+}
+
+// MerkleizeVector uses our optimized routine to hash a list of 32-byte
+// elements.
+func merkleizeVector(elements [][32]byte, length uint64) [32]byte {
+	dep := depth(length)
+	// Return zerohash at depth
+	if len(elements) == 0 {
+		return zeroHashesRaw[dep]
+	}
+	for i := uint8(0); i < dep; i++ {
+		layerLen := len(elements)
+		oddNodeLength := layerLen%2 == 1
+		if oddNodeLength {
+			zerohash := zeroHashesRaw[i]
+			elements = append(elements, zerohash)
+		}
+		outputLen := len(elements) / 2
+		err := gohashtree.Hash(elements, elements)
```

Comment:
Perhaps adding comment here that gohashtree overwrites concurrently elements is worth it.